### DISCUSSION TITLE
dotnet 6 minimal api with api gateway - How do I get the APIGatewayProxyRequest ?

### DISCUSSION THREAD

**Discussion by michaelakin** [aws/aws-lambda-dotnet#1292] [Q&A] [answered]
Hi,

I have a dot net 6 minimal api lambda with the following.  how can I get the APIGatewayProxyRequest stage variables from api gateway?

```
app.MapGet("/test", ([FromQuery] string name, [FromServices] APIGatewayProxyRequest req) =>
{
    Log.Information("Name {name}", name);

    return name;
}).WithName("Test");
```

I am getting the following error with this set up:
```
[23:16:42 ERR] Unknown error responding to request: InvalidOperationException:
System.InvalidOperationException: No service for type 'Amazon.Lambda.APIGatewayEvents.APIGatewayProxyRequest' has been registered.
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at lambda_method3(Closure , Object , HttpContext )
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass36_0.<Create>b__0(HttpContext httpContext)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)
--- End of stack trace from previous location ---
   at Amazon.Lambda.AspNetCoreServer.AbstractAspNetCoreFunction`2.ProcessRequest(ILambdaContext lambdaContext, Object context, InvokeFeatures features, Boolean 
```

**Comment by normj** [comment, marked as answer]
If you are goal is to access the `APIGatewayProxyRequest` object you can access it from the `HttpContext.Items` collection. Here is an example.

```
app.MapGet("/test", ([FromQuery] string name, HttpContext context, [FromServices] ILogger<Program> logger) =>
{
    
    if(context.Items.TryGetValue(Amazon.Lambda.AspNetCoreServer.AbstractAspNetCoreFunction.LAMBDA_REQUEST_OBJECT, out var req))
    {
        var lambdaRequest = req as APIGatewayProxyRequest;
        logger.LogInformation($"Found lambda request: {lambdaRequest != null}");
    }
    else
    {
        logger.LogInformation("No lambda request found, probably running it locally outside of Lambda");
    }

    return name;
}).WithName("Test");
```

We have talked about adding extension methods to the `HttpContext` to make it more obvious how to get a hold of the object.

**Reply by michaelakin** [reply]
Thanks, I did find that and was able to access it, but we thought surely there was a better and easier way.  I guess I will make my own extension methods for now.

**Reply by michaelakin** [reply]
also @normj   is this what we would have to do for other events like s3, or sqs etc?

**Reply by michaelakin** [reply]
@normj  ?? ^^

**Reply by normj** [reply]
@michaelakin For other event sources like S3 you wouldn't be using ASP.NET Core so no. The ASP.NET Core bridge library is just for API Gateway and Application Load Balancer events. So to write an S3 Lambda function you would use the basic Lambda programming model having a function that takes in an S3Event. 

You also might take a look at our new [Amazon.Lambda.Annotations](https://github.com/aws/aws-lambda-dotnet/tree/master/Libraries/src/Amazon.Lambda.Annotations) library which integrates .NET dependency injection and CloudFormation syncing. Most of the examples are still focused on API Gateway use cases but you can use it other types of events as well to take advantage of the DI setup and CloudFormation syncing.
